### ISSUE TITLE
Restrict RHS compound set visibility by target AND target access string

### ISSUE THREAD

**Issue by mwinokan** [m2ms/fragalysis-frontend#1552]
E.g. currently in staging there are two versions of `A71EV2A`, and one under `lb32627-66` and the other under `lb32627-71`. 

I uploaded a RHS dataset to `lb32627-66` but it also shows up under `lb32627-71`:
![Screenshot 2024-10-22 at 14 54 10](https://github.com/user-attachments/assets/de379b05-df81-4316-9bdc-34e7cb073ea9)

@matej-vavrek can you please check that the f/e requests only the compound sets with the correct target name AND target access string/proposal?

**Comment by matej-vavrek** [comment]
@mwinokan, it seems that we are fetching target via `api/targets/?title=A71EV2A` . @kaliif but I see in API filter options only `title` so adding access string like for example `api/targets/?title=A71EV2A&target_access_string=lb18145-1` probably will not have any effect or?

**Comment by kaliif** [comment]
@matej-vavrek adding TAS to filter fields is not a problem, but even without it, the backend should not return data from other proposals. This is possibly more a backend issue than a frontend.

**Comment by kaliif** [comment]
Looking at this issue a bit more closely, @matej-vavrek why do you need to look up target here? `api/compound-sets/` endpoint allows you to filter by target ID, if you have that available, you can fetch the correct compound set. @mwinokan if you can confirm that you are a member in both proposals and there's no data leak, I think that would solve the problem.

I can of course add additional fields to filters if that makes anyone's life more comfortable.

**Comment by mwinokan** [comment]
Thanks @kaliif, indeed the same target name under different proposals should be treated as a separate project with its own compound sets, etc. 

I noticed that the same kind of issue exists with SiteObservationTags, but I concluded that this was actually a useful behaviour

**Comment by matej-vavrek** [comment]
> Looking at this issue a bit more closely, @matej-vavrek why do you need to look up target here? `api/compound-sets/` endpoint allows you to filter by target ID, if you have that available, you can fetch the correct compound set. @mwinokan if you can confirm that you are a member in both proposals and there's no data leak, I think that would solve the problem.
> 
> I can of course add additional fields to filters if that makes anyone's life more comfortable.

I am not sure if I understand correctly @kaliif. When is some target choosed from the target list at the main page we get that target and then it loads the other data.

**Comment by mwinokan** [comment]
Included in b/e [PR687](https://github.com/xchem/fragalysis-backend/pull/687)

**Comment by mwinokan** [comment]
Moving back to in progress because RHS compound sets are visible across the same target under different proposals in staging

**Comment by mwinokan** [comment]
Both the [66](https://fragalysis.xchem.diamond.ac.uk/viewer/react/preview/target/A71EV2A/tas/lb32627-66) and [71](https://fragalysis.xchem.diamond.ac.uk/viewer/react/preview/target/A71EV2A/tas/lb32627-71) versions are using target=35 to fetch from `api/compound-sets/` so the compound sets are the same

**Comment by matej-vavrek** [comment]
@mwinokan, I have updated [my stack](https://fragalysis-matej-default.xchem-dev.diamond.ac.uk/). It should use target name and access string to get a target now, e.g.
`https://fragalysis-matej-default.xchem-dev.diamond.ac.uk/api/targets/?title=A71EV2A&project=2`

**Comment by mwinokan** [comment]
I have tested it on @matej-vavrek's stack. It's not fully working.

1. I uploaded two versions of the `A71EV2A_xca_staging_20241104.tar.gz` under lb32627-66 and lb32627-71
2. I uploaded [A71EV2A_lhs2rhs_20241010_capitalprefix_first3.sdf.zip](https://github.com/user-attachments/files/17904838/A71EV2A_lhs2rhs_20241010_capitalprefix_first3.sdf.zip) to 66. The generated name was: `A71EV2A_lhs2rhs_2024-2024-11-25-A`
3. I then checked both 66 and 71 targets and the behaviour was correct, I could only see the RHS compound set for target 66
4. I uploaded [A71EV2A_lhs2rhs_20241010_capitalprefix.sdf.zip](https://github.com/user-attachments/files/17904862/A71EV2A_lhs2rhs_20241010_capitalprefix.sdf.zip) to 71. The generated name was **again**: `A71EV2A_lhs2rhs_2024-2024-11-25-A`
5. The behaviour in the RHS of both targets **did not change**. I was just seeing the one RHS compound set in 66 and none in 71.

My guess would be that since the generated compound set name is the same, it might be a pk clash and no second compound set entry is created, 66 fetches the original one fine, but 71 sees no compound sets with the correct TAS.

I suspect this may be one for @kaliif

**Comment by mwinokan** [comment]
@phraenquex asks how the b/e can be returning data from a different TAS, should it not be blocked by authentication?

@kaliif says it's likely because @mwinokan is authorised for both

@kaliif confirms that the compound set name is being used as a primary key, and this is likely the problem. @kaliif please let @mwinokan know if a new spec is needed for compound set names

**Comment by kaliif** [comment]
From the backend perspective, the correct solution would be to add a new primary key and demote the compound set name from the primary key to a text field. @matej-vavrek would that work in the FE? You'd have to fetch the sets using the new ID or by name+target combination (assuming we won't allow duplicate names within a target @mwinokan?).

**Comment by mwinokan** [comment]
That sounds sensible @kaliif. Indeed the compound set name should be unique within a Target.

That reminds me, when using the `upload_cset` endpoint the options in the `update/delete existing set` drop-down, is it possible to add the target name (and TAS) to make it clearer what you are deleting?

**Comment by kaliif** [comment]
> That reminds me, when using the `upload_cset` endpoint the options in the `update/delete existing set` drop-down, is it possible to add the target name (and TAS) to make it clearer what you are deleting?

@mwinokan  yes, but as I'm not proficient with Vue.js, I can't give you a reasonable estimate of how difficult this will be.

**Comment by matej-vavrek** [comment]
> From the backend perspective, the correct solution would be to add a new primary key and demote the compound set name from the primary key to a text field. @matej-vavrek would that work in the FE? You'd have to fetch the sets using the new ID or by name+target combination (assuming we won't allow duplicate names within a target @mwinokan?).

I think if you tell us where and what to add it should work. Sadly, I am not able to test it locally, because we have only one public target access string.

**Comment by matej-vavrek** [comment]
@kaliif thank you! I have noticed  `PUBLIC_TAS`, but was not sure if it could be extended to more. I will try it.

**Comment by mwinokan** [comment]
@kaliif reports that this was quite a complicated migration which has been achieved, but he is double checking the upload functionality

**Comment by kaliif** [comment]
I'm done with the backend. There were quite a few changes in many areas of the code and there may be something I missed. I suggest, before merging, to test this a bit. I've deployed it to my stack. @mwinokan let me know if you disagree, I can create a PR.
@matej-vavrek the changes that affect the FE - there's no more `api/computedset-download` endpoint, I moved the download to `api/compound-sets` where it makes more sense. The download page is `api/compound-sets/<id>/download`. And everywhere the FE fetches the compound sets, it now needs to happen by numeric ID, not name. Filtering using target and project is available.
In addition to my stack, the backend image is also available in dockerhub - `kaliif/fragalysis-backend:latest`

**Comment by kaliif** [comment]
Found and fixed. @mwinokan the backend should now allow to fetch compound sets with ID: https://fragalysis-kalev-default.xchem-dev.diamond.ac.uk/api/compound-sets/1/.
@matej-vavrek the code is deployed to my stack and there's also an image, `kaliif/fragalysis-backend:latest` if you prefer to test locally.
There may be similar issues coming up, compound sets are used in multiple places and changing the foreign key is bound to break things. So don't be alarmed when it crashes, just let me know.

**Comment by mwinokan** [comment]
@matej-vavrek please check the status of the frontend development for this ticket

**Comment by mwinokan** [comment]
@matej-vavrek is loading the latest backend image and will see how much work is needed for the frontend

**Comment by mwinokan** [comment]
@kaliif says this ticket contains a large migration that needs merging. 

@matej-vavrek please check the f/e status ASAP and then deploy to your stack so @mwinokan can test

**Comment by mwinokan** [comment]
@matej-vavrek the upload of two different RHS compoud sets to the same target but under two different TAS seems to have worked, but the f/e for the RHS is crashing the page due to a 400 error from: https://fragalysis-matej-default.xchem-dev.diamond.ac.uk/api/compound-mols-scores/?computed_set=Fstein_Thumb-2025-01-30-A. I think you need to include the TAS in the request

**Comment by mwinokan** [comment]
Thanks @matej-vavrek. I can confirm the behaviour on the RHS is now as I expect. Please merge it to staging

**Comment by matej-vavrek** [comment]
@mwinokan, [PR](https://github.com/xchem/fragalysis-frontend/pull/476) merged into staging

**Comment by mwinokan** [comment]
All looks fine in staging
